sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,264 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ComponentModel;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System
+{
+    /// <summary>
+    /// A lightweight abstraction for a payload of bytes that supports converting between string, stream, JSON, and bytes.
+    /// </summary>
+    public class BinaryData
+    {
+        private const int CopyToBufferSize = 81920;
+
+        /// <summary>
+        /// The backing store for the <see cref="BinaryData"/> instance.
+        /// </summary>
+        private readonly ReadOnlyMemory<byte> _bytes;
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the
+        /// provided byte array.
+        /// </summary>
+        /// <param name="data">The array to wrap.</param>
+        public BinaryData(byte[] data)
+        {
+            _bytes = data;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object to JSON
+        /// using <see cref="JsonSerializer"/>.
+        /// </summary>
+        ///
+        /// <param name="jsonSerializable">The object that will be serialized to JSON using
+        /// <see cref="JsonSerializer"/>.</param>
+        /// <param name="options">The options to use when serializing to JSON.</param>
+        /// <param name="type">The type to use when serializing the data. If not specified, <see cref="object.GetType"/> will
+        /// be used to determine the type.</param>
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+        public BinaryData(object jsonSerializable, JsonSerializerOptions? options = default, Type? type = default)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+        {
+            _bytes = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, type, options);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the
+        /// provided bytes.
+        /// </summary>
+        /// <param name="data">Byte data to wrap.</param>
+        public BinaryData(ReadOnlyMemory<byte> data)
+        {
+            _bytes = data;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from a string by converting
+        /// the string to bytes using the UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        public BinaryData(string data)
+        {
+            _bytes = Encoding.UTF8.GetBytes(data);
```

Comment:
Encoding.UTF8.GetBytes is going to throw an ArgumentNullException for the parameter "s".  It should be manually validated here.

```suggestion
            if (data == null)
                throw new ArgumentNullException(nameof(data));

            _bytes = Encoding.UTF8.GetBytes(data);
```

Since the FromString method uses the same parameter name it can just defer to this one.